llama_index/tools/function_tool.py

```diff
@@ -14,7 +15,8 @@ def sync_to_async(fn: Callable[..., Any]) -> AsyncCallable:
     """Sync to async."""
 
     async def _async_wrapped_fn(*args: Any, **kwargs: Any) -> Any:
-        return fn(*args, **kwargs)
+        loop = asyncio.get_running_loop()
+        return await loop.run_in_executor(None, lambda: fn(*args, **kwargs))
```

Comment:
This is still going to block though right? The `fn` is not actually async or using `await` within it.

The existing implementation is at least a little clearer that it will be blocking no?

Reply:
Hmm maybe I take that back. I ran your test script with this fix

```
Added user message to memory: What is 11 + 22?
[test2] waiting 0 2024-01-05 19:50:13.444879
[test2] waiting 1 2024-01-05 19:50:14.447270
=== Calling Function ===
Calling function: add with args: {
  "a": 11,
  "b": 22
}
[add] waiting 0 2024-01-05 19:50:14.839473
[test2] waiting 2 2024-01-05 19:50:15.448229
[add] waiting 1 2024-01-05 19:50:15.839699
[test2] waiting 3 2024-01-05 19:50:16.449476
[add] waiting 2 2024-01-05 19:50:16.839950
[test2] waiting 4 2024-01-05 19:50:17.450745
Got output: 33
========================

The sum of 11 and 22 is 33.
[test2] waiting 5 2024-01-05 19:50:18.452306
[test2] waiting 6 2024-01-05 19:50:19.453549
[test2] waiting 7 2024-01-05 19:50:20.454799
[test2] waiting 8 2024-01-05 19:50:21.456050
[test2] waiting 9 2024-01-05 19:50:22.457338
```

Reply:
It's allowing the two functions to alternate running, due to the top-level await handing over the event loop. 

But the coroutine itself will still be blocking the entire event loop since it doesn't call `await` internally

Reply:
Can we leave a comment that this still means that the tool itself will block the event loop when it gets ran? Like this won't magically make it non-blocking

Reply:
Dear @logan-markewich,

Thank you for your thorough review and feedback. I wanted to address your concerns regarding the potential blocking of the event loop. While it's true that the `fn` function itself does not contain an `await` statement, the `loop.run_in_executor` wrapper ensures that `fn` executes in a separate background thread. This means that it does not block the event loop while it runs. This ensures that operations like database access or network requests behave similarly to `time.sleep(1)` in `add`, without blocking the event loop. The event loop remains free to handle other async tasks, providing the intended asynchronous behavior. The test script's output also supports this, as it shows the functions alternating execution without blocking.

Best regards.